p2panda-rs/src/storage_provider/traits/entry_store.rs

```diff
@@ -0,0 +1,295 @@
+        &self,
+        author: &Author,
+        log_id: &LogId,
+        seq_num: &SeqNum,
+    ) -> Result<Option<StorageEntry>, EntryStorageError>;
+
+    /// Returns the latest Bamboo entry of an author's log.
+    async fn latest_entry(
+        &self,
+        author: &Author,
+        log_id: &LogId,
+    ) -> Result<Option<StorageEntry>, EntryStorageError>;
+
+    /// Return vector of all entries of a given schema
+    async fn by_schema(&self, schema: &SchemaId) -> Result<Vec<StorageEntry>, EntryStorageError>;
+
+    /// Determine skiplink entry hash ("lipmaa"-link) for entry in this log, return `None` when no
+    /// skiplink is required for the next entry.
+    async fn determine_skiplink(
+        &self,
+        entry: &StorageEntry,
+    ) -> Result<Option<Hash>, EntryStorageError> {
+        let next_seq_num = entry.entry_decoded().seq_num().clone().next().unwrap();
+
+        // Unwrap as we know that an skiplink exists as soon as previous entry is given
+        let skiplink_seq_num = next_seq_num.skiplink_seq_num().unwrap();
+
+        // Check if skiplink is required and return hash if so
+        let entry_skiplink_hash = if is_lipmaa_required(next_seq_num.as_u64()) {
+            let skiplink_entry = match self
+                .entry_at_seq_num(
+                    &entry.entry_encoded().author(),
+                    entry.entry_decoded().log_id(),
+                    &skiplink_seq_num,
+                )
+                .await?
+            {
+                Some(entry) => Ok(entry),
+                None => Err(EntryStorageError::SkiplinkMissing),
+            }?;
+            Ok(Some(skiplink_entry.entry_encoded().hash()))
+        } else {
+            Ok(None)
+        }?;
+
+        Ok(entry_skiplink_hash)
+    }
+}
+
+#[cfg(test)]
+pub mod tests {
+
+    use async_trait::async_trait;
+    use rstest::rstest;
+    use std::sync::{Arc, Mutex};
+
+    use crate::entry::{sign_and_encode, Entry, EntrySigned, LogId, SeqNum};
+    use crate::identity::{Author, KeyPair};
+    use crate::operation::{AsOperation, OperationEncoded};
+    use crate::schema::SchemaId;
+    use crate::storage_provider::errors::EntryStorageError;
+    use crate::storage_provider::traits::test_setup::{
+        test_db, SimplestStorageProvider, StorageEntry, SKIPLINK_ENTRIES,
+    };
+    use crate::storage_provider::traits::{AsStorageEntry, EntryStore};
+    use crate::test_utils::fixtures::{
+        entry, entry_signed_encoded, operation_encoded, random_key_pair, schema,
+    };
+
+    /// Implement `EntryStore` trait on `SimplestStorageProvider`
+    #[async_trait]
+    impl EntryStore<StorageEntry> for SimplestStorageProvider {
+        /// Insert an entry into storage.
+        async fn insert_entry(&self, entry: StorageEntry) -> Result<bool, EntryStorageError> {
+            let mut entries = self.entries.lock().unwrap();
+            entries.push(entry);
+            // Remove duplicate entries.
+            entries.dedup();
```

Comment:
yeh, true, this is a bit weird even for a test impl... Not sure quite what you mean by putting it into a new method, but I'd rather just swap out the vec based storage for a hashmap where duplicates would just overwrite or some kind of set where duplicates would fail. How does that sound to you?

Reply:
What do you think of this test setup in general? I know we will have an in-memory `StorageProvider` implementation, but for the tests I feel it's better to have something like this which is very simple and transparent. Detached from the rest of the code base.

Reply:
I would have introduced a method in the storage provider implementation wrapping that functionality:

```rust
Impl SProvider {
  Pub fn insert(entry) {
    Self.push
    Self.dedupe
  }
}
```

HashMaps are great ofc as well! Thought you had a reason to use vecs 👍

Reply:
Ah, ok, cool, yes that's nice. I didn't have a very special reason for vecs, I was having fun representing all the data as strings, then thought just pushing the in a vec was the simplest storage I could think of. ;-p